File: DFS, BFS/11724.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, m, result;
bool visited[1001];
vector<int> graph[1001];

void dfs(int x){
    visited[x] = true;
    for(int i = 0; i < graph[x].size(); i++){
        int now = graph[x][i];
        if(!visited[now]) dfs(now);
    }
}

int main(void){
    cin >> n >> m;
    for(int i = 0; i < m; i++){
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    for(int i = 1; i <= n; i++){
        if(!visited[i]){
            dfs(i);
            result++;
        }
    }
    cout << result << '\n';
    return 0;
}

File: DFS, BFS/11724.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

n, m = map(int, input().split())

graph = [[] for _ in range(n + 1)]
visited = [False] * (n + 1)

for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

for i in range(1, n + 1):
    graph[i].sort()

def dfs(graph, v, visited):
    visited[v] = True

    for i in graph[v]:
        if not visited[i]:
            dfs(graph, i, visited)

cnt = 0
for i in range(1, n + 1):
    if visited[i] == False:
        cnt += 1
        dfs(graph, i, visited)

print(cnt)
